Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
+
+Correct and effective code generation for structs requires that the JIT have ready access
+to information about the shape and size of the struct. This information is obtained from the VM
+over the JIT/EE interface. This includes:
+- Struct size
+- Number and type of fields, especially if there are GC references
+
+It has been proposed by @mikedn in
+[#21705 Pull struct type info out of GenTreeObj](https://github.com/dotnet/coreclr/pull/21705)
+that this kind of information could be kept in a `ClassLayout` cache, reducing the size impact on the `GenTree`
+nodes, as well as the number
+of JIT/EE interface queries. Reducing the impact on the size of the nodes would make it less
+costly to retain this information even on rvalues, thus enabling some
+optimization scenarios that are currently disabled for struct types.
+
+Current Representation of Struct Values
+---------------------------------------
+### Importer-Only Struct Values
+
+These struct-typed nodes are created by the importer, but transformed in morph, and so are not
+encountered by most phases of the JIT:
+* `GT_INDEX`: This is transformed to a `GT_IND`
+* `GT_FIELD`: This is transformed to a `GT_LCL_FLD` or a `GT_IND`
+* `GT_MKREFANY`: This produces a "known" struct type, which is currently obtained by
+  calling `impGetRefAnyClass()` which is a call over the JIT/EE interface. This node is always
+  eliminated, and its source address used to create a copy. If it is on the rhs
+  of an assignment, it will be eliminated during the importer. If it is a call argument it will
+  be eliminated during morph.
+  * The presence of any of these in a method disables struct promotion. See `case CEE_MKREFANY` in the
+    `Importer`, where it is asserted that these are rare, and therefore not worth the trouble to handle.
+
+### Struct “objects” as lvalues
+
+* The lhs of a struct assignment is a block node or local
+  * `GT_OBJ` nodes represent the “shape” info via a struct handle, along with the GC info
+    (location and type of GC references within the struct).
+    * These are used only to represent struct values that contain GC references.
+  * `GT_BLK` nodes represent struct types with no GC references, or opaque blocks of fixed size.
+    * These have no struct handle, resulting in some pessimization or even incorrect
+      code when the appropriate struct handle can't be determined.
+    * These never represent lvalues of structs that contain GC references.
+    * Proposed: Like `GT_OBJ`, these would have an index into the `ClassLayout` cache.
+  * `GT_STORE_OBJ` and `GT_STORE_BLK` have the same structure as `GT_OBJ` and `GT_BLK`, respectively
+    * `Data()` is op2
+  * `GT_DYN_BLK` and `GT_STORE_DYN_BLK` (GenTreeDynBlk extends GenTreeBlk)
+    * Additional child `gtDynamicSize`
+    * Note that these aren't really struct types; they represent dynamically sized blocks
+      of arbitrary data.
+  * For `GT_LCL_VAR` nodes, the shape information is obtained via the `LclVarDsc`
```

Comment:
Mention `LCL_FLD` as well? `ClassLayout` was started with the goal of allowing `LCL_FLD` struct typed nodes.

Reply:
LclFld with layout support: https://github.com/dotnet/coreclr/pull/21711/commits/aa086e4a67a9c56eb95185b4552592bdfcaf006f